x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/SearchableSnapshotIndexInput.java

```diff
@@ -121,9 +121,19 @@ private void readInternalBytes(final int part, long pos, final byte[] b, int off
             // we did not read everything in an optimized fashion, so read the remainder directly
             try (InputStream inputStream
                      = blobContainer.readBlob(fileInfo.partName(part), pos + optimizedReadSize, length - optimizedReadSize)) {
-                final int directReadSize = inputStream.read(b, offset + optimizedReadSize, length - optimizedReadSize);
+
+                int directReadSize = 0;
+                while ((optimizedReadSize + directReadSize) < length) {
+                    final int read = inputStream.read(b,
+                                                      offset + optimizedReadSize + directReadSize,
+                                                      length - optimizedReadSize - directReadSize);
+                    if (read == -1) {
+                        break;
```

Comment:
This means we hit the end of the file, which I think is unexpected. In tests the assertion below will fail; in production we will proceed even though some of the buffer contains its old contents, which would be bad. I think instead we should throw an `EOFException` here.

Reply:
> I think instead we should throw an `EOFException` here.

I pushed e06a94b0f558ed2b85fdf1975856141a1d42210e